### DISCUSSION TITLE
[Idea] Allow modules to integrate their own bundles

### DISCUSSION THREAD

**Discussion by jolelievre** [PrestaShop/PrestaShop#35078] [Ideas]
### The problem

Today the modules can integrate in the core via:
- using hooks
- using legacy controller
- defining their own Symfony controllers and Symfony services
- overriding/decorating the Symfony services from the core
- I prefer not to talk about the legacy Override feature since it's not recommended

But there's one thing the modules can't do: add their own Symfony bundles.

This would give lots of new possibilities:
- create and handle their module code via a modern custom bundle instead of relying on the Module class for many tech features
- integrate Symfony bundles that are not in the core to add new technical features
- add other compiler passes that are required in some cases to improve DI or to optimize the container build
- with the bundle comes a Symfony extension and with it the possibility to adapt the framework configuration, so the modules could finally change the project configuration without having to modify the core files

### The solution

The solution would target the v9.0, which is being upgraded to Symfony 6.0. The registration of the bundle is a bit different in recent Symfony versions, we'd need to adapt our `AppKernel::registerBundles` a bit but now the bundles are defined in a separated PHP file like this:

```php
// config/bundles.php
return [
    // 'all' means that the bundle is enabled for any Symfony environment
    Symfony\Bundle\FrameworkBundle\FrameworkBundle::class => ['all' => true],
    Symfony\Bundle\SecurityBundle\SecurityBundle::class => ['all' => true],
    Symfony\Bundle\TwigBundle\TwigBundle::class => ['all' => true],
    Symfony\Bundle\MonologBundle\MonologBundle::class => ['all' => true],
    Doctrine\Bundle\DoctrineBundle\DoctrineBundle::class => ['all' => true],
    Sensio\Bundle\FrameworkExtraBundle\SensioFrameworkExtraBundle::class => ['all' => true],
    // this bundle is enabled only in 'dev' and 'test', so you can't use it in 'prod'
    Symfony\Bundle\WebProfilerBundle\WebProfilerBundle::class => ['dev' => true, 'test' => true],
];
```

Usually, this file is handled by flex but we don't use it yet in the core, but that's not the issue here. The principle is that the list of bundles can returned as a simple PHP file that contains an associative array of bundles.

We could have the same file in the modules (for example in `modules/my_module/bundles.php`), our `AppKernel::registerBundles` method could loop through all the **installed modules** and search for this file, if present it merges its content with the on from the core, we just need to prevent duplicate values and handle the merge of associated environments for each bundle.

And that's it.

The implementation itself is quite simple, however of course it would require testing it's impact that can't all be anticipated in this discussion. It also means that we open more freedom to the modules with this feature, so they must be careful not to mess too much with the core, but it's already possible to heavily mess with the core services, so I'm not sure the move is that risky. It all comes down to the module code quality in the end.

### Do you plan to work on this subject?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by matks** [comment]
Related
https://github.com/PrestaShop/PrestaShop/issues/14541
https://github.com/PrestaShop/PrestaShop/issues/27188
